packages/forms/src/form_builder.ts

```diff
@@ -20,162 +20,196 @@ function isAbstractControlOptions(options: AbstractControlOptions|
+    *
+    * @usageNotes
+    *
+    * ### Initialize a control as disabled
+    *
+    * The following example returns a control with an initial value in a disabled state.
+    *
+    * <code-example path="forms/ts/formBuilder/form_builder_example.ts" region="disabled-control">
+    * </code-example>
+    */
+   control<T = any>(
+       formState: T|FormState<T>,
+       validatorOrOpts?: ValidatorFn|ValidatorFn[]|FormControlOptions|null,
+       asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|
+       null): FormControl<T|null>|FormControl<T> {
+     return new FormControl(formState, validatorOrOpts, asyncValidator);
+   }
+
+   /**
+    * Constructs a new `FormArray` from the given array of configurations,
+    * validators and options.
+    *
+    * @param cc An array of child controls or control configs. Each
+    * child control is given an index when it is registered.
+    *
+    * @param validatorOrOpts A synchronous validator function, or an array of
+    * such functions, or an `AbstractControlOptions` object that contains
+    * validation functions and a validation trigger.
+    *
+    * @param asyncValidator A single async validator or array of async validator
+    * functions.
+    */
+   array<T = any>(
+       childControls: Array<T|FormState<T>|ControlConfig<T>|AbstractControl<T>|FormControl<T>>,
+       validatorOrOpts?: ValidatorFn|ValidatorFn[]|AbstractControlOptions|null,
+       asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|
+       null): FormArray<AbstractControl<T|null>> {
+     const controls = childControls.map(c => this._createControl(c));
+     return new FormArray(controls, validatorOrOpts, asyncValidator);
+   }
+
+   /** @internal */
+   _reduceControls<T>(childControls:
+                          {[k: string]: T|ControlConfig<T>|FormState<T>|AbstractControl<T>}):
+       {[key: string]: AbstractControl} {
+     const controls: {[key: string]: AbstractControl} = {};
+     Object.keys(childControls).forEach(controlName => {
+       controls[controlName] = this._createControl(childControls[controlName]);
+     });
+     return controls;
+   }
+
+   /** @internal */
+   _createControl<T>(childControls: T|FormState<T>|ControlConfig<T>|AbstractControl<T>|
+                     FormControl<T>): AbstractControl<T>|AbstractControl<T|null> {
+     if (childControls instanceof FormControl) {
+       return childControls as FormControl<T>;
+     } else if (childControls instanceof AbstractControl) {  // A control; just return it
+       return childControls;
+
+     } else if (Array.isArray(childControls)) {  // ControlConfig Array
+       const value: T|FormState<T> = childControls[0];
+       const validator: ValidatorFn|ValidatorFn[]|null =
+           childControls.length > 1 ? childControls[1]! : null;
+       const asyncValidator: AsyncValidatorFn|AsyncValidatorFn[]|null =
+           childControls.length > 2 ? childControls[2]! : null;
+       return this.control<T>(value, validator, asyncValidator);
```

Comment:
All type annotations and the non-null assertion operators are unnecessary here, as TS has narrowed `childControls` to be the tuple `ControlConfig` here, because that's the only array type among the constituents of the `childControls` parameter's union type. But feel free to keep them if you think it helps clarify the code.

```suggestion
       const value = childControls[0];
       const validator = childControls.length > 1 ? childControls[1] : null;
       const asyncValidator = childControls.length > 2 ? childControls[2] : null;
       return this.control<T>(value, validator, asyncValidator);
```

Reply:
Hmm that's surprising, IIRC I did try the code I suggested and I don't remember seeing any errors. But happy to keep as is as this is indeed non-trivial to follow.